### PR TITLE
layout, sync to `skeleton`, start fixing network egs

### PR DESCRIPTION
**Pull request by mor1**
work to this point on OSX for me. yet to test linux.

Do not merge yet. Also relies on mirage/mirage-skeleton#98

### CODE REVIEW

**Comment by mor1** [comment]
Ok-- I think this one is ready to be merged assuming Travis goes green.
@amirmc @avsm @djs55 @dsheets @samoht Any chance someone could quickly review please?

File: tmpl/wiki/hello-world.md [hunk 1]
```diff
@@ -106,21 +108,19 @@ the [documentation](http://mirage.github.io) for this module.
...
 ```
 
-`mirage configure` will generate a `Makefile` with all the build rules included
-from evaluating the configuration file.  It also creates a `main.ml` that
-represents the entry point of your unikernel.  `make depend` will first check
-that you have all the right OPAM packages installed to build a Unix
-application, and will install them if they're not present.
+`mirage configure` generates a `Makefile` with all the build rules included from
+evaluating the configuration file, and a `main.ml` that represents the entry
+point of your unikernel. The `configure` step should ensure all external nd OPAM
```
Comment (Drup) [inline tmpl/wiki/hello-world.md]: Extra `nd` on this line.

File: tmpl/wiki/hello-world.md [hunk 2]
```diff
@@ -425,121 +661,284 @@ let net =
...
+    href='https://github.com/mirage/mirage-http'>HTTP library</a>.
+  </p>
 </div>
 
-Let's get the network stack compiling on Unix first.  On a Mac, be sure
-to install the [tuntap](http://tuntaposx.sourceforge.net/) kernel module
+#### Unix / Socket networking
+
+Let's get the network stack compiling on Unix first. On a pre-Yosemite Mac, be
+sure to install the [tuntap](http://tuntaposx.sourceforge.net/) kernel module
```
Comment (Drup) [inline tmpl/wiki/hello-world.md]: Maybe find a non-sourceforge link ? :)

File: tmpl/wiki/hello-world.md [hunk 3]
```diff
@@ -425,121 +661,284 @@ let net =
...
+`vmnet` framework replaces `tuntap`._
+
+By default, if we do not use DHCP with a `direct` network stack, Mirage will
+configure the stack to use the `tap0` interface with an address of `10.0.0.2`.
+This requires the tuntap bridge to be configured after the unikernel starts to
+ensure we have a route to it. Thus:
+
+```
+$ cd stackv4
+$ env NET=direct DHCP=true mirage configure --unix
```
Comment (Drup) [inline tmpl/wiki/hello-world.md]: Shouldn't it be DHCP=false ?

**Comment by mor1** [comment]
Thanks! All fixed except the sourceforge link -- sourceforge appears to be in static mode currently, and I'm not aware of a proper homepage for tuntap other than sourceforge. If you've got one, happy to add it... :/

File: tmpl/wiki/deployment.md [hunk 4]
```diff
@@ -1,140 +1,155 @@
...
+source code on [mirage/mirage-www](https://github.com/mirage/mirage-www) on
+GitHub. Our workflow is such that we can send a
+[pull request](https://github.com/mirage/mirage-www/pulls?direction=desc&page=1&sort=created&state=closed)
+to update the website, and have a fully standalone Xen kernel committed into the
+binary
+[mirage/mirage-www-deployment](https://github.com/mirage/mirage-www-deployment)
+repository.
 
 This all works by installing a fresh MirageOS installation for each change
 committed to the website, and rebuilding the entire website kernel from scratch
```
Comment (amirmc) [inline tmpl/wiki/deployment.md]: kernel -> unikernel

File: tmpl/wiki/deployment.md [hunk 5]
```diff
@@ -1,140 +1,155 @@
...
+${TMP_BUILD}/travis-mirage
+```
+
+This fetches a helper library,
+[`yorick`](https://github.com/ocaml/ocaml-travisci-skeleton/blob/master/yorick.mli),
+and builds it; and finally fetches, builds and executes
+[travis_mirage.ml](https://github.com/ocaml/ocaml-travisci-skeleton/blob/master/travis_mirage.ml)
+with the environment variables defined in the
+[.travis.yml](https://github.com/mirage/mirage-www/.travis.yml) file. This piece
+of OCaml sets the environment to use OPAM, installs any requested pins, and then
```
Comment (amirmc) [inline tmpl/wiki/deployment.md]: Where do the requested pins go?

File: tmpl/wiki/deployment.md [hunk 6]
```diff
@@ -146,64 +161,90 @@ a small `travis-senv` command-line tool to make this easier.
...
-While this fragment is a bit longer than the rest, it's all rather
-straightforward.  We install `travis-senv` and use it to retrieve
-our SSH deployment key (which is stored encrypted in the `.travis.yml`
-file, and can only be decrypted by the Travis infrastructure).
-Once that's installed, we initialize `git` and clone the deployment
-repository.  We then compress our output kernel and commit it to
-the repository along with the configuration data.
+(The `?|` and `?|~` operators are defined in [`yorick.ml`](https://github.com/ocaml/ocaml-travisci-skeleton/blob/master/yorick.ml) and act simply as shell
+invocations piping `stdin`/`stdout`/`stderr` in the usual way, with
+success/failure tests on the status of each invoked command.)
```
Comment (amirmc) [inline tmpl/wiki/deployment.md]: I appreciate this explanation :)

File: tmpl/wiki/deployment.md [hunk 7]
```diff
@@ -146,64 +161,90 @@ a small `travis-senv` command-line tool to make this easier.
...
-repository for new pushes, and updates to the latest version in
-that repository when one appears.  This machine doesn't need to
-have any of the MirageOS tools installed, and if there's an error in
-an updated website, we can simply revert back to a previous changeset
-of the deployed site.
+The live website runs on a Xen machine which has a crontab entry in dom0 that
+polls the
+[mirage/mirage-www-deployment](http://github.com/mirage/mirage-www-deployment)
+repository for new content via `git pull`. When new content __is__ obtained, a
+Git [`merge-hook`](https://github.com/mirage/mirage-www-deployment) is invoked
```
Comment (amirmc) [inline tmpl/wiki/deployment.md]: `merge-hook` -> `post-merge` hook

Maybe explicitly reference the scripts here and link to them in the deploy repo? There may be folks who aren't as familiar with git's hooks so I wonder if linking to the manual would be useful (http://git-scm.com/docs/githooks#_post_merge)

File: tmpl/wiki/deployment.md [hunk 8]
```diff
@@ -146,64 +161,90 @@ a small `travis-senv` command-line tool to make this easier.
...
+there's an error in an updated website, we can simply revert back to a previous
+as previous binary revisions of the site remain in the deployment repository.
 
 The same approach could be used to deploy to Heroku, Rackspace or EC2 instead of
 a custom Xen host; we'll cover this in future updates.
 
-*(Note: This article is based on a similar blog post for general OCaml use available [here](http://anil.recoil.org/2013/10/06/travis-secure-ssh-integration.html))*
+*(Note: This article is based on a similar blog post for general OCaml use
+ available
+ [here](http://anil.recoil.org/2013/10/06/travis-secure-ssh-integration.html).)*
```
Comment (amirmc) [inline tmpl/wiki/deployment.md]: The following use specific examples to guide the user through details of what the scripts are doing — though the scripts themselves re general-purpose. .

- http://amirchaudhry.com/from-jekyll-to-unikernel-in-fifty-lines/
- http://amirchaudhry.com/heroku-for-unikernels-pt1/

File: tmpl/wiki/hello-world.md [hunk 9]
```diff
@@ -142,131 +154,287 @@ If you are on a 64-bit Linux system able to build Xen images, simply change
...
+# disk = [ '/dev/loop0,,xvda' ]
+
+# The network configuration is defined here:
+# http://xenbits.xen.org/docs/4.3-testing/misc/xl-network-configuration.html
+# An example would look like:
+# vif = [ 'mac=c0:ff:ee:c0:ff:ee,bridge=br0' ]
+```
+
+Edit this to customise the VM name or memory, and then run it (if you're still
+on an old release of Xen you may need to swap the `xl` command for `xm`):
```
Comment (amirmc) [inline tmpl/wiki/hello-world.md]: Good to know but is it possible to list a version number for 'old' Xen?

File: tmpl/wiki/hello-world.md [hunk 10]
```diff
@@ -142,131 +154,287 @@ If you are on a 64-bit Linux system able to build Xen images, simply change
...
+world
+$
+```
+
+As you can see, after some initial boot messages have been displayed as Xen
+boots the VM, you see the same output on the Xen console as you did on the UNIX
+version you ran earlier. If you need more help, or would like to see how to boot
+your Xen VM on Amazon's EC2, [click here](/wiki/xen-boot). Finally, if you are
+using an ARM processor, such as the CubieBoard, you will need some extra OPAM
+package pins that can be found
```
Comment (amirmc) [inline tmpl/wiki/hello-world.md]: Are they still pins?

File: tmpl/wiki/hello-world.md [hunk 11]
```diff
@@ -297,46 +467,98 @@ let disk1 = crunch "t"
...
+YES!
+YES!
+YES!
+YES!
+YES!
+YES!
+YES!
+```
+
+Of course, this scheme doesn't really scale up to large website, and we often
```
Comment (amirmc) [inline tmpl/wiki/hello-world.md]: website -> websites

**Comment by Drup** [comment]
For sourceforce, I don't sorry. 

I like the improvements on the hello-world tutorial. 

The mirage-www is better, but I'm still dubious about its point, as said on the mailing list. It explains how to build and play with the website, not how to build your own, and if you want to delve into the details, it doesn't help you. 

I can try everything on linux tomorrow, if you need it.

File: tmpl/wiki/hello-world.md (hunk 9 above)
Comment (mor1) [inline tmpl/wiki/hello-world.md]: Sure, but I don't know the answer. 
@djs55 @avsm ?

Comment (djs55) [inline tmpl/wiki/hello-world.md]: xl became default in xen-4.2 and xm was removed in xen-4.5. I would suggest: use xm if xend is running and xl otherwise.

**Comment by mor1** [comment]
@amirmc thanks, all fixed I think.
@drup thanks; i've now tested on linux on virtualbox so hopefully ok. (but there's no such thing as too much testing, right? :)

**Comment by mor1** [comment]
@djs55 thanks, tried to clarify

**Pull request merged**